src/Pact/Native.hs

```diff
@@ -157,18 +157,28 @@ enforceOneDef =
   ["(enforce-one \"Should succeed on second test\" [(enforce false \"Skip me\") (enforce (= (+ 2 2) 4) \"Chaos reigns\")])"]
   "Run TESTS in order (in pure context, plus keyset enforces). If all fail, fail transaction. Short-circuits on first success."
   where
+    enforceBool i e = reduce e >>= \case
+      tl@(TLiteral (LBool cond) _) -> if cond then pure (Just tl) else pure Nothing
+      _ -> evalError' i "Failure: expected boolean result in enforce-one"
+    catchTxFailure :: PactError -> Eval e (Maybe (Term Name))
+    catchTxFailure e = case peType e of
+      TxFailure -> pure Nothing
+      _ -> throwM e
     enforceOne :: NativeFun e
     enforceOne i as@[msg,TList conds _ _] = runReadOnly i $
       gasUnreduced i as $ do
         msg' <- reduce msg >>= \case
           TLitString s -> return s
           _ -> argsError' i as
         let tryCond r@Just {} _ = return r
-            tryCond Nothing cond = catch
-              (Just <$> reduce cond)
-              -- TODO: instead of catching all here, make pure violations
-              -- independently catchable
-              (\(_ :: SomeException) -> return Nothing)
+            tryCond Nothing cond = isExecutionFlagSet FlagDisablePact44 >>= \case
+              True -> do
+                g <- getGas
+                catch (Just <$> reduce cond)
+                  -- TODO: instead of catching all here, make pure violations
+                  -- independently catchable
+                  (\(_ :: SomeException) -> putGas g *> return Nothing)
```

Comment:
TODO: separate pr makes sense here and we can redo the pact error stuff in one swoop